Add ``vqa_dataset``, update docs

krammnic opened the pull request:
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
#1704

#### Changelog
What are the changes made in this PR?
* #1704

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [X] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [X] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [X] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [X] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [X] I have added an example to docs or docstrings

krammnic commented:
@RdoubleA Require review.

krammnic commented:
I see where is the problem. I committed and then switched branch...

krammnic commented:
@RdoubleA Should be fine now.

RdoubleA commented on torchtune/datasets/multimodal/_multimodal_instruct.py:
```diff
@@ -0,0 +1,158 @@
+    *,
+    source: str,
+    column_map: Optional[Dict[str, str]] = None,
+    train_on_input: bool = False,
```
we are removing train_on_input for multimodal datasets, as it is not really used. you'll notice that `multimodal_chat_dataset` also has this removed

RdoubleA commented on torchtune/datasets/multimodal/_multimodal_instruct.py:
```diff
@@ -0,0 +1,158 @@
+    column_map: Optional[Dict[str, str]] = None,
+    train_on_input: bool = False,
+    new_system_prompt: Optional[str] = None,
+    packed: bool = False,
```
we don't currently support packing for multimodal (yet... we are planning to add this soon)

krammnic replied:
@RdoubleA Is there will be a problem to add packing in multimodal? There is no issue about it as I see. Might open then next PR with it

RdoubleA replied:
Yes, there is non trivial work to ensure the cross attention masks are also packed and collated correctly, so it would involve 1) adding logic to pack the cross attention masks and 2) create a new collated for packed multimodal

We are actively discussing an overhaul to our dataloader / collate / packing logic in the near term, so I would hold off on multimodal packing support for now.

krammnic commented:
Yeah, some stuff failed( Let me fix then other PRs and then this one.

RdoubleA commented on torchtune/datasets/multimodal/_multimodal_instruct.py:
```diff
@@ -0,0 +1,158 @@
+    new_system_prompt: Optional[str] = None,
+    packed: bool = False,
+    filter_fn: Optional[Callable] = None,
+    split: str = "train",
```
would also be good to include the `image_dir` functionality similar to multimodal chat: https://github.com/pytorch/torchtune/blob/main/torchtune/datasets/multimodal/_multimodal.py#L22. this would enable users to configure datasets that either have image paths or the images directly. For example, take a look a MathVista (https://huggingface.co/datasets/AI4Math/MathVista), a math image QA benchmark that has both the image paths and the raw image. some datasets may have one or the other.

This may require some non-trivial changes to InputOutputToMessages however to make it support loading images. I'm happy to provide more guidance on that, but it would be similar to the changes to `ShareGPTToMessages` in this pr: https://github.com/pytorch/torchtune/pull/1667

RdoubleA replied:
nvm, I see you already made those changes :)

RdoubleA commented on torchtune/datasets/multimodal/_multimodal_instruct.py:
```diff
@@ -0,0 +1,158 @@
+    **load_dataset_kwargs: Dict[str, Any],
+) -> Union[SFTDataset, PackedDataset]:
+    """
+    Configure a custom multimodal dataset with user instruction prompts and model responses.
```
the dataset builders say multimodal but this is specifically text+image so would be good to call it out. Maybe something like:
```suggestion
    Configure a custom text+image dataset with separate columns for user instruction, image, and model response.
```

RdoubleA commented on torchtune/datasets/multimodal/_multimodal_instruct.py:
```diff
@@ -0,0 +1,158 @@
+
+        |  input          |  image          |  output          |
+        |-----------------|-----------------|------------------|
+        | "user prompt"   | "user image"    | "model response" |
```
```suggestion
        | "user prompt"   | images/1.jpg    | "model response" |
```
It should be an image path, or the image itself (not sure how to represent that in a text example). saying just "user image" is not clear what data type should go here

RdoubleA commented on torchtune/datasets/multimodal/_multimodal_instruct.py:
```diff
@@ -0,0 +1,158 @@
+
+        column_map = {"input": "question", "output": "answer", "image": "picture"}
+
+    Masking of the prompt during training is controlled by the ``train_on_input`` flag, which is
```
remove train_on_input

RdoubleA commented on torchtune/datasets/multimodal/_multimodal_instruct.py:
```diff
@@ -0,0 +1,158 @@
+
+
+def multimodal_instruct_dataset(
+    tokenizer: ModelTokenizer,
```
for multimodal datasets, this needs to be:
```suggestion
    model_transform: Transform,
```
I would follow `multimodal_chat_dataset` more closely instead of `instruct_dataset` for this one

RdoubleA commented on torchtune/datasets/multimodal/_multimodal_instruct.py:
```diff
@@ -0,0 +1,158 @@
+    ::
+
+        >>> from torchtune.datasets.multimodal import multimodal_instruct_dataset
+        >>> dataset = multimodal_instruct_dataset(
```
maybe structure it similar to this example: https://github.com/pytorch/torchtune/blob/7744608c4c455a86b21f4ce0642e6c19bc1cebf4/torchtune/datasets/multimodal/_multimodal.py#L123

and FlamingoTransform should be replaced everywhere in the example with `Llama3VisionTransform` https://github.com/pytorch/torchtune/blob/main/torchtune/models/llama3_2_vision/_transform.py#L17

RdoubleA commented on torchtune/data/_messages.py:
```diff
@@ -161,20 +161,23 @@ class InputOutputToMessages(Transform):
         train_on_input: bool = False,
         column_map: Optional[Dict[str, str]] = None,
         new_system_prompt: Optional[str] = None,
+        is_multimodal: bool = False,
```
can we infer this similar to in ShareGPTToMessages?
```
is_multimodal = "image" in sample or (
            "image" in self._column_map and self._column_map["image"] in sample
        )
```
so users don't need to worry about another flag

krammnic replied:
Sure!

RdoubleA reviewed (commented):
Could you test this builder with an example VQA dataset once comments are addressed? Something like https://huggingface.co/datasets/derek-thomas/ScienceQA using the question and solution columns.

If you're able to easily configure this dataset using the builder you added, and you can successfully decode tokens of a random sample, and the iamge is encoded correctly, then I would declare this builder to be well-tested (in addition to the unit test you put up)

krammnic commented on tests/assets/multimodal_instruct_tiny.json:
```diff
@@ -0,0 +1,7 @@
+[
+    {
+        "input": "What is presented on image?",
+        "output": "PyTorch logo.",
```
Yeah, that was example

ebsmothers commented on tests/torchtune/datasets/multimodal/test_multimodal_instruct_dataset.py:
```diff
@@ -0,0 +1,58 @@
+        ]
+        
+        expected_labels = [
+            [np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(7), np.int64(5), np.int64(-1)]
```
nit but is the `np.int64` strictly necessary? I know labels are long dtype, but if we are just comparing raw lists via `==` and not using torch tensor asserts or something I think you shouldn't need it (and it makes the expected values easier to read)

krammnic replied:
Yeah, will be fixed.

krammnic commented:
Done all small fixes + CI fixes. Only need extra test

krammnic commented:
@RdoubleA @ebsmothers 
Made all required fixes, manually tested on required dataset:
```
from torchtune.datasets.multimodal import multimodal_instruct_dataset
import torch
from torch import nn
from torchtune.data import Message, PromptTemplate, truncate
from torchtune.modules.tokenizers import ModelTokenizer
from torchtune.modules.transforms import Transform
from typing import Any, Dict, Generator, List, Mapping, Optional, TextIO, Tuple, Union


class DummyTokenizer(ModelTokenizer, Transform):
    def __init__(self, max_seq_len: Optional[int] = None):
        self.max_seq_len = max_seq_len

    def encode(self, text, add_bos=True, add_eos=True, **kwargs) -> List[int]:
        words = text.split()
        tokens = [len(word) for word in words]
        if add_bos:
            tokens = [self.bos_id] + tokens
        if add_eos:
            tokens = tokens + [self.eos_id]
        return tokens

    def tokenize_messages(
        self,
        messages: List[Message],
    ) -> Tuple[List[int], List[bool]]:
        """
        A simplified version of Llama2Tokenizer's ``tokenize_messages`` for testing purposes.
        """
        start_of_turn = True
        end_of_turn = False
        tokenized_messages = []
        mask = []
        for message in messages:
            # If assistant message, this is the end of a turn
            end_of_turn = message.role == "assistant"

            # Prepend BOS on start of new turns
            if start_of_turn:
                tokenized_messages.append(self.bos_id)
                mask.append(message.masked)

            # Tokenize current message, append with masks
            tokens = []
            for item in message.content:
                if item["type"] == "text":
                    tokens = tokens + self.encode(
                        item["content"],
                        add_bos=False,
                        add_eos=False,
                    )
                elif item["type"] == "image":
                    tokens = tokens + [self.image_id]

            tokenized_messages.extend(tokens)
            mask.extend([message.masked] * len(tokens))

            # If assistant message, append EOS at end
            if end_of_turn:
                tokenized_messages.append(self.eos_id)
                mask.append(message.masked)
                end_of_turn = False
                start_of_turn = True
            else:
                start_of_turn = False

            # Break out early if we reach max_seq_len
            if self.max_seq_len and len(tokenized_messages) >= self.max_seq_len:
                break

        # Finally, truncate if necessary
        if self.max_seq_len:
            tokenized_messages = truncate(
                tokenized_messages, self.max_seq_len, self.eos_id
            )
            mask = truncate(mask, self.max_seq_len, message.masked)

        return tokenized_messages, mask

    def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
        messages: List[Message] = sample.pop("messages")
        images = []
        for message in messages:
            images += message.get_media()
        tokens, mask = self.tokenize_messages(messages)
        sample["tokens"] = tokens
        sample["mask"] = mask
        sample["images"] = images
        return sample

    @property
    def eos_id(self):
        return -1

    @property
    def bos_id(self):
        return 0

    @property
    def image_id(self):
        return -2

dataset = multimodal_instruct_dataset(
    model_transform=DummyTokenizer(),
    # source="json",
    source="derek-thomas/ScienceQA",
    # data_files="derek-thomas/ScienceQA",
    column_map={
        "input": "question",
        "output": "solution",
        "image": "image"
    },
    split="train"
)

tokens = dataset[0]
print(tokens)
```
Output:

```
{'input': 'question', 'output': 'solution', 'image': 'image'}
{'tokens': [0, -2, 5, 2, 5, 6, 2, 8, 6, 2, 4, 3, 7, 4, 2, 3, 7, 5, 4, 2, 5, 3, 3, 5, 5, 2, 9, 4, 8, 2, 8, 6, -1], 'mask': [True, True, True, True, True, True, True, True, True, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False], 'images': [<PIL.PngImagePlugin.PngImageFile image mode=RGB size=750x429 at 0x7FD20A590D50>], 'labels': [np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(-100), np.int64(2), np.int64(4), np.int64(3), np.int64(7), np.int64(4), np.int64(2), np.int64(3), np.int64(7), np.int64(5), np.int64(4), np.int64(2), np.int64(5), np.int64(3), np.int64(3), np.int64(5), np.int64(5), np.int64(2), np.int64(9), np.int64(4), np.int64(8), np.int64(2), np.int64(8), np.int64(6), np.int64(-1)]}

```

I'm pretty confident that it is actually working

krammnic commented:
@ebsmothers @RdoubleA Added another fix, locally tests are passed

krammnic commented:
Yeah, finnaly. Can someone approve and merge? Or should we do some extra checks?

ebsmothers commented on torchtune/data/_messages.py:
```diff
@@ -175,30 +175,64 @@ def __init__(
-            self._column_map = {"input": "input", "output": "output"}
+            self.column_map = {"input": "input", "output": "output"}
+
+        self._column_map = column_map
```
Duplicate of L179?

ebsmothers commented on torchtune/data/_messages.py:
```diff
@@ -175,30 +175,64 @@ def __init__(
+        if self.column_map:
+            if "input" not in self.column_map:
                 raise ValueError(
-                    f"Expected a key of 'input' in column_map but found {column_map.keys()}."
+                    f"Expected a key of 'input' in column_map but found {self.column_map.keys()}."
                 )
-            if "output" not in column_map:
+            if "output" not in self.column_map:
                 raise ValueError(
-                    f"Expected a key of 'output' in column_map but found {column_map.keys()}."
+                    f"Expected a key of 'output' in column_map but found {self.column_map.keys()}."
                 )
-            self._column_map = column_map
         else:
-            self._column_map = {"input": "input", "output": "output"}
+            self.column_map = {"input": "input", "output": "output"}
+
+        self._column_map = column_map
```
This feels duplicative.. why do we need to define both `self.column_map` and `self._column_map`?

krammnic replied:
Hmm, will think about this, problem that we can't fully define column_mao in `__init__` because of is_multimodal check in`__call__`

ebsmothers commented on torchtune/datasets/multimodal/_multimodal_instruct.py:
```diff
@@ -0,0 +1,140 @@
+    filter_fn: Optional[Callable] = None,
+    split: str = "train",
+    **load_dataset_kwargs: Dict[str, Any],
+) -> Union[SFTDataset, PackedDataset]:
```
Strictly speaking if we don't support packing yet should return type just be `SFTDataset`?

krammnic replied:
Sure, typehint needs to be fixed

ebsmothers commented on tests/torchtune/datasets/multimodal/test_multimodal_instruct_dataset.py:
```diff
@@ -0,0 +1,52 @@
+            )
+            assert prompt == expected_tokens[i]
+            assert label == expected_labels[i]
+            assert isinstance(image[0], PngImageFile) is True
```
nit: do we need `is True` here?

krammnic replied:
Probably True might be removed, yes

krammnic commented on tests/torchtune/datasets/multimodal/test_multimodal_instruct_dataset.py:
```diff
@@ -0,0 +1,52 @@
+    def tokenizer(self):
+        return DummyTokenizer()
+
+    @pytest.mark.parametrize("train_on_input", [True, False])
+    def test_get_item(self, tokenizer, train_on_input):
```
Yeah, good point

krammnic commented:
@ebsmothers Fixed, according to your comments

krammnic commented:
Can someone start CI and merge then?

joecummings commented on torchtune/data/_messages.py:
```diff
@@ -175,30 +175,60 @@ def __init__(
+
+        self.column_map = column_map
+
+        if self.column_map:
```
```suggestion
        if self.column_map is not None:
```

joecummings commented on torchtune/datasets/multimodal/_multimodal_instruct.py:
```diff
@@ -0,0 +1,139 @@
+          split: train
+
+    Returns:
+        Union[SFTDataset, PackedDataset]: the configured :class:`~torchtune.datasets.SFTDataset`
```
Return type here needs to be fixed.

joecummings commented on torchtune/data/_messages.py:
```diff
@@ -175,30 +175,60 @@ def __init__(
+            self.column_map = {"input": "input", "output": "output"}
 
     def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
+        if self.column_map is not None:
```
I don' think we can make the assumption that "image" will always be the column for media. I'd much rather have the `ToMessages` classes take care of converting any images to the PIL format. 

Can you remove this section from this PR?

Then, we can think about how to modify our message converters to handle this loading.

RdoubleA replied:
> I don' think we can make the assumption that "image" will always be the column for media.

we are not making this assumption, users can still change the "image" column in column_map, in which case we pull from there

joecummings replied:
This should be handled in a `ToMessage` formatter, not here though.

RdoubleA replied:
this is in `InputOutputToMessages` which seems sensible to me, were you thinking of a different location?

SalmanMohammadi replied:
I'm with @RdoubleA on this, `InputOutputToMessages` handles this. We should be documenting this, though.

joecummings reviewed (commented):
Mostly looks good, but I don't think the Message class should be automatically converting the images to PIL b/c it's assuming which column is the image input.

RdoubleA commented:
hey @krammnic , I've added you on Discord. Let's coordinate further there with @joecummings and we can update this thread after

joecummings commented on torchtune/datasets/multimodal/_vqa.py:
```diff
@@ -0,0 +1,138 @@
+    .. code-block:: yaml
+
+        dataset:
+          _component_: torchtune.datasets.multimodal.multimodal_instruct_dataset
```
Update.

krammnic replied:
Fixed

RdoubleA commented on tests/assets/multimodal_instruct_tiny.json:
nit: can rename this to vqa_tiny.json

RdoubleA commented on torchtune/data/_messages.py:
```diff
@@ -175,30 +175,60 @@ def __init__(
     def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
+        if self.column_map is not None:
+            is_multimodal = "image" in sample or (
+                "image" in self.column_map and self.column_map["image"] in sample
```
if "image" is in sample, we can assume the user is not using `column_map`. The `or` should short circuit and return true without checking column_map, so we don't need to check `if self.column_map is not None:` above. we can also add "image" key in default column map:
`self.column_map = {"input": "input", "output": "output", "image": "image"}`

RdoubleA commented on torchtune/data/_messages.py:
```diff
@@ -175,30 +175,60 @@ def __init__(
+        else:
+            is_multimodal = "image" in sample
+
+        if is_multimodal and "image" not in self.column_map:
```
if we use the default column_map in my comment above if user does not pass in column map, then we don't need to do this

RdoubleA commented on torchtune/datasets/multimodal/_vqa.py:
```diff
@@ -0,0 +1,138 @@
+    **load_dataset_kwargs: Dict[str, Any],
+) -> SFTDataset:
+    """
+    Configure a custom text+image dataset with separate columns for user instruction, image, and model response.
```
Can we specify this docstring more for visual question-answer?

RdoubleA commented on torchtune/datasets/multimodal/_vqa.py:
```diff
@@ -0,0 +1,138 @@
+
+    .. code-block:: text
+
+        |  input          |  image          |  output          |
```
same with this example. I wonder if we should use quesiton and answer as default columns instead of input and output, but I'm ok either way

RdoubleA commented on torchtune/datasets/multimodal/_vqa.py:
```diff
@@ -0,0 +1,138 @@
+
+        column_map = {"input": "question", "output": "answer", "image": "picture"}
+
+    Masking of the prompt during training is controlled by the ``train_on_input`` flag, which is
```
since we've removed train_on_input, we should remove this from the docstring

RdoubleA commented on torchtune/datasets/multimodal/_vqa.py:
```diff
@@ -0,0 +1,138 @@
+    new_system_prompt: Optional[str] = None,
+    filter_fn: Optional[Callable] = None,
+    split: str = "train",
+    **load_dataset_kwargs: Dict[str, Any],
```
do we need an image_dir parameter if users have downloaded the iamges to a particular local directory?

RdoubleA replied:
bumping this comment

RdoubleA commented on torchtune/data/_messages.py:
```diff
@@ -175,30 +175,60 @@ def __init__(
+            image_path = sample[self.column_map["image"]]
+            if isinstance(image_path, str):
+                # Load if not loaded
+                pil_image = load_image(image_path)
```
the images themselves could be saved to some local directory if the user had to download the image dataset separately. we should provide an option to pass a path to prepend to all image paths in the dataset in this case

RdoubleA reviewed (commented):
Thanks for the VQA rename we discussed offline and all your work so far. We should also update the rst docs to make sure this builder is discoverable there. There's a few comments that I think should be addressed but we're nearly there.

krammnic commented:
Pushed some fixes

krammnic commented:
I have a feeling that we will have to separate 2 version of InputToMessages to not break SRP for example

krammnic commented:
@RdoubleA @joecummings Can we merge?

SalmanMohammadi commented on torchtune/data/_messages.py:
```diff
@@ -175,30 +175,54 @@ def __init__(
     ):
         self.train_on_input = train_on_input
```
I have this slight nagging feeling that this transform is over-generalized and it wouldn't be harmful to define another transform which should properly document that the column map can include image, and to also include validation logic, and without needing to check for multimodal.

Fine to leave this as a followup but if there's consensus here I'd like to see an issue.

SalmanMohammadi replied:
I see this is the same for `ShareGPTToMessages`, so maybe it's fine. We need to properly document MM usage either way.

RdoubleA replied:
Yeah, I see what you mean. Especially as we add more modalities this may get bloated quick. I was the one who initiated this design in ShareGPTToMessages but I think it's worth reconsidering. I'll put up an issue.

SalmanMohammadi commented on torchtune/datasets/multimodal/_vqa.py:
```diff
@@ -0,0 +1,132 @@
+          data_files: my_dataset.json
+          column_map:
+            input: question
+            output: answer
```
```suggestion
            output: answer
            image: picture
```

SalmanMohammadi commented on torchtune/datasets/multimodal/_vqa.py:
```diff
@@ -0,0 +1,132 @@
+          column_map:
+            input: question
+            output: answer
+          train_on_input: False
+          packed: False
+          split: train
```
```suggestion
          split: train
```

SalmanMohammadi commented on torchtune/datasets/multimodal/_vqa.py:
```diff
@@ -0,0 +1,132 @@
+        column_map = {"input": "question", "output": "answer", "image": "picture"}
+
+    Args:
+        model_transform (Transform): callable that applies model-specific pre-processing to the sample.
```
Can we be more concrete here? For example, the cauldron dataset builder describes this arg as:
```python
        model_transform (Transform): model-specific transform class that takes in a sample dict and applies custom
            transforms on the keys. It should consist of at minimum two components: text tokenization (called
            on the "messages" field) and image transform (called on the "images" field). The keys returned by
            the model transform should be aligned with the expected inputs into the model.
```

SalmanMohammadi commented on torchtune/datasets/multimodal/_vqa.py:
```diff
@@ -0,0 +1,132 @@
+        | "user prompt"   | images/1.jpg    | "model response" |
+
+    If your column names are different, you can use the ``column_map`` parameter to change
+    the expected column names. For example, if your dataset has columns ``"question"``,
+    ``"answer"`` and ``"picture"`` you can use:
+
+        column_map = {"input": "question", "output": "answer", "image": "picture"}
```
I'm not sure if we need a concrete example of using column_map here, it is better described below.

krammnic replied:
I'm not sure if it is. For me it looks fine, but we can remove this description if you want

SalmanMohammadi commented on torchtune/datasets/multimodal/_vqa.py:
```diff
@@ -0,0 +1,132 @@
+            and "output" column names to the actual column names in the dataset. Keys should be "input" and
+            "output" and values should be the actual column names. Default is None, keeping the default "input"
+            and "output" column names.
+        new_system_prompt (Optional[str]): if specified, prepend a system message. This can
+            serve as instructions to guide the model response. Default is None.
```
```suggestion
        new_system_prompt (Optional[str]): if specified, prepend a system message. This can
            serve as instructions to guide the model response. Setting this will OVERRIDE any system
            messages already present in the dataset. Default is None.
```
We can be more descriptive here - borrowing from mm chat builder.

SalmanMohammadi commented:
This looks great to me. Coming at this from an outside perspective I'd really love to see a couple follow up issues to make sure we properly document new features we add to our datasets. I've mentioned above some things I immediately found confusing when combing through the current docs in this PR. 

- [ ] We now can document usage of a general vqa MM builder in https://pytorch.org/torchtune/main/basics/multimodal_datasets.html, to the same extent which we have documented the MM chat builder. Alternatively, we can rename the current MM docpage to `Multimodal chat datasets`, and add a new page for MM instruct.
- [ ] We need to add a concrete HF builder using this, and add it at the bottom of https://pytorch.org/torchtune/main/basics/multimodal_datasets.html (or a new instruct MM page)
- [ ] I'd like to see a fine-tuning run with this dataset.

krammnic commented:
> This looks great to me. Coming at this from an outside perspective I'd really love to see a couple follow up issues to make sure we properly document new features we add to our datasets. I've mentioned above some things I immediately found confusing when combing through the current docs in this PR.
> 
>     * [ ]  We now can document usage of a general vqa MM builder in https://pytorch.org/torchtune/main/basics/multimodal_datasets.html, to the same extent which we have documented the MM chat builder. Alternatively, we can rename the current MM docpage to `Multimodal chat datasets`, and add a new page for MM instruct.
> Probably the second idea is better.
>     * [ ]  We need to add a concrete HF builder using this, and add it at the bottom of https://pytorch.org/torchtune/main/basics/multimodal_datasets.html (or a new instruct MM page)
> 
>     * [ ]  I'd like to see a fine-tuning run using this dataset.
> Then will run full procedure

Should I make this docs fixes in this PR or open separate?

krammnic commented:
Oh, lint(I accepted changes right here). Let me fix

SalmanMohammadi commented:
> Then will run full procedure
> Should I make this docs fixes in this PR or open separate?

We'd be incredibly grateful if you're up for helping out here! I was speaking slightly generally just to make sure some of those tasks get done at some point, so myself (and the other maintainers) are happy to help out, particularly with a training run, and of course, reviews!

krammnic commented:
Some really weird lint fail.

krammnic commented:
Is there anything else to add in this PR?

RdoubleA commented on torchtune/datasets/multimodal/_vqa.py:
```diff
@@ -0,0 +1,135 @@
+            in the filepath in ``data_files``, and set ``split="train"``. See `Hugging Face's
+            <https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path>`_
+            ``load_dataset`` for more details.
+        image_dir (str): path to the directory containing the images as you are expected to download the COCO dataset
```
this should be generalized and not specific to COCO. Maybe use the same docstring from image_dir argument in `InputOutputToMessages`?

krammnic replied:
Fixed

RdoubleA reviewed (approved):
Appreciate you pushing this all the way through. I think this highlighted some rough parts on our multimodal pipeline which we should follow up on and created some productive discussions. Just one comment, otherwise it's good to go once CI is green.
